addons/hr_timesheet/hr_timesheet.py

```diff
@@ -103,129 +77,118 @@ def on_change_unit_amount(self, cr, uid, id, prod_id, unit_amount, company_id, u
+        if emp and emp[0].journal_id.id:
+            return emp[0].journal_id.id
+        else:
+            raise exceptions.ValidationError("This user or employee is not associated to a valid Journal ID")
+
+    is_timesheet = fields.Boolean()
+   
+    @api.constrains('user_id')
+    def check_user_id(self):
+        if self.is_timesheet:
+            emp = self.env['hr.employee'].search([('user_id','=',self.user_id.id or self.env.uid)])
+            if not emp:
+                raise exceptions.ValidationError("There is no employee defined for user " + self.user_id.name)
+            else:
+                if not emp[0].journal_id.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Analytic Journal. Please define one for him or select another user.")
+                elif not emp[0].product_id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Product. Please define one for him or select another user.")
+                elif not emp[0].product_id.uom_id.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Product Unit of Measure. Please define one for him or select another user.")
+                elif not emp[0].product_id.categ_id.property_account_expense_categ.id:
+                    raise exceptions.ValidationError("The employee " + emp[0].name + " is not associated to a valid Financial Account. Please define one for him or select another user.")
+
+    @api.onchange('user_id')
+    def V8_on_change_user_id(self):
+        if self.is_timesheet:
+            new_values = self.on_change_user_id(self.user_id.id, self.is_timesheet)
+            self.journal_id = new_values["value"]['journal_id']
+            self.product_id = new_values["value"]['product_id']
+            self.product_uom_id = new_values["value"]['uom_id']
+            self.general_account_id = new_values["value"]['general_account_id']
+
+            # New API style
+            #     emp = self.env['hr.employee'].search([('user_id','=',self.user_id.id or self.env.uid)])
+            #     if not emp:
+            #         model, action_id = self.env['ir.model.data'].get_object_reference('hr', 'open_view_employee_list_my')
+            #         msg = _("Employee is not created for this user. Please create one from configuration panel.")
+            #         raise exceptions.RedirectWarning(msg, action_id, _('Go to the configuration panel'))
+            #     else:
+            #         self.journal_id = self._get_analytic_journal()
+            #         self.product_id = self._get_employee_product()
+            #         self.product_uom_id = self._get_employee_unit()
+            #         self.general_account_id = self._get_general_account()
+
+    def on_change_user_id(self, cr, uid, ids, user_id, is_timesheet=False, context=None):
+        if is_timesheet:
```

Comment:
You did not check the else condition, or a default return. An onchange should always return at least a void dict.